aws-lc-rs/src/rsa/encryption.rs

```diff
@@ -0,0 +1,359 @@
+            UsageContext::Encryption,
+        )?))
+    }
+
+    /// Decrypts the contents in `ciphertext` and writes the corresponding plaintext to `output`.
+    ///
+    /// # Sizing `output`
+    /// For `OAEP_SHA1_MGF1SHA1`, `OAEP_SHA256_MGF1SHA256`, `OAEP_SHA384_MGF1SHA384`, `OAEP_SHA512_MGF1SHA512` a safe
+    /// `output` length is the key size in bytes. The RSA key size in bytes can be retrieved using [`Self::key_size`].
+    /// The actual output length will at most `rsaKeySizeBytes - (2 * hashLengthBytes) - 2`.
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs while decrypting `ciphertext`.
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs while decrypting `ciphertext`.
+    pub fn decrypt<'output>(
+        &self,
+        algorithm: &'static EncryptionAlgorithm,
+        ciphertext: &[u8],
+        output: &'output mut [u8],
+    ) -> Result<&'output mut [u8], Unspecified> {
+        let pkey_ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.0.key, null_mut()) })?;
+
+        if 1 != unsafe { EVP_PKEY_decrypt_init(*pkey_ctx) } {
+            return Err(Unspecified);
+        }
+
+        match algorithm {
+            EncryptionAlgorithm::OAEP(oaep) => {
+                configure_oaep_crypto_operation(
+                    &pkey_ctx,
+                    oaep.oaep_hash_fn(),
+                    oaep.mgf1_hash_fn(),
+                )?;
+            }
+        }
+
+        let mut out_len = output.len();
+
+        if 1 != indicator_check!(unsafe {
+            EVP_PKEY_decrypt(
+                *pkey_ctx,
+                output.as_mut_ptr(),
+                &mut out_len,
+                ciphertext.as_ptr(),
+                ciphertext.len(),
+            )
+        }) {
+            return Err(Unspecified);
+        };
+
+        Ok(&mut output[..out_len])
+    }
+}
+
+impl Debug for PrivateDecryptingKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("PrivateDecryptingKey").finish()
+    }
+}
+
+impl AsDer<Pkcs8V1Der<'static>> for PrivateDecryptingKey {
+    fn as_der(&self) -> Result<Pkcs8V1Der<'static>, Unspecified> {
+        AsDer::<Pkcs8V1Der<'_>>::as_der(&self.0)
+    }
+}
+
+/// An RSA public key used for encrypting plaintext that is decrypted by a [`PrivateDecryptingKey`].
+pub struct PublicEncryptingKey(RsaEvpPkey);
+
+impl PublicEncryptingKey {
+    /// Construct a `PublicEncryptingKey` from X.509 `SubjectPublicKeyInfo` DER encoded bytes.
+    ///
+    /// # Errors
+    /// * `Unspecified` for any error that occurs deserializing from bytes.
+    pub fn from_der(value: &[u8]) -> Result<PublicEncryptingKey, KeyRejected> {
+        Ok(Self(RsaEvpPkey::from_rfc5280_public_key_der(
+            value,
+            UsageContext::Encryption,
+        )?))
+    }
+
+    /// Returns the RSA key size in bytes.
+    #[must_use]
+    pub fn key_size(&self) -> usize {
+        self.0.key_size()
+    }
+
```

Comment:
We should provide a utility function for consumers:
```
pub fn max_plaintext_size(&self, algorithm: &'static EncryptionAlgorithm) -> usize { ... }
```

I don't think we need one for ciphertext size as long as the documentation is clear that it matches the key size.